accepted/exception-handling.md

```diff
@@ -0,0 +1,514 @@
+experiments on top of Wasmtime.
+
+Therefore we propose to start by implementing exceptions as
+uncatchable traps. This will force implementation of the initial
+`wasmtime::Config` knobs, CLI flags, some Wasm-to-CLIF translation,
+etc... It allows toolchains, like Kotlin, to start running subsets of
+their test suite on Wasmtime, even if they use exceptions-related
+instructions, as long as the dynamic execution doesn't actually throw
+any exceptions.
+
+This is also a mode that we may want to support indefinitely anyways:
+some programs "must" be compiled with exceptions due to accidental
+complexity with (say) their build system, but won't ever actually
+raise any exceptions due to the configuration of various knobs. With
+this mode, Wasmtime can run these programs and do so without the
+`.cwasm` binary bloat associatred with the exceptions' side tables or
+the save-all-registers execution overhead of `try_call`.
+
+Finally, it is worth noting that this mode is not visible to Wasm
+guests in any way, and is therefore fully spec compliant (the same
+way, e.g., Wasmtime's fuel feature is).
+
+## Unwinding Across Instances
+[unwinding-instances]: #unwinding-across-instances
+
+It is worth noting that we must support unwinding across frames from
+different modules and instances. For example, if one instance imports
+and calls an exported function from a different instance, and then
+that function raises an exception:
+
+```
+instance_A -> instance_B -> raise exception E
+```
+
+Note that the original caller, `instance_A`, can install an exception
+handler for `E` and catch the exception that `instance_B` raises.
+
+Our outlined implementation approach supports this scenario out of the
+box. We mention it here only for completeness.
+
+## Host APIs
+[host-apis]: #host-apis
+
+As stated in the [non-requirements] section, we do not plan to support
+unwinding across host frames. However, we must allow and provide APIs
+for host functions to
+
+* inspect, catch, and propagate exceptions raised by now-unwound Wasm
+  callees, and
+* raise new exceptions to pass up the stack to their Wasm callers.
+
+We propose to have our host-to-wasm trampolines catch all exceptions
+raised by Wasm and return them to the host via a three-way result type
+that is logically an `enum` of either a normal successful return
+value, an exception, or a trap.
+
+Similarly, we will do not plan to supporting raising Wasm exceptions
+and beginning unwinding directly in the middle of arbitrary host
+code. Instead, we envisage a host function will return the exception
+variant of this three-way result type, potentially early-exiting via
+`?`-propagation. When returning to Wasm from the host, we will unpack
+that three-way result, check for the exception variant and, if found,
+enter the unwinder instead of continuing with either normal
+control-flow returns or the jump-directly-to-the-trap-handler
+`longjmp` path.[^setjmp-longjmp]
```

Comment:
To confirm, the thinking here is that "longjmp" is a standin for "restore the stack and jump to a pc", probably via an assembly stub, right? Basically (a) not `longjmp` itself, (b) not a cranelift-synthesized stub, and (c) simple enough that we can maintain handwritten assembly

Reply:
No, sorry if this isn't clear: this sentence is referring to the existing `longjmp`s we have for panics and traps that skip over Wasm code. That `longjmp` path is mentioned just as a path we would _skip_ if there is an exception (instead of a panic/trap that we need to propagate over Wasm frames).

Eventually we want to have inline Cranelift stubs to replace that, but that is orthogonal to this RFC, as mentioned in the footnote.